activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -114,12 +114,16 @@ def join_constraints(outer_joins, join_type)
           end
         }
 
-        joins.concat outer_joins.flat_map { |oj|
+        joins.concat joins_to_add.flat_map { |oj|
           if join_root.match? oj.join_root
             walk join_root, oj.join_root
           else
             oj.join_root.children.flat_map { |child|
-              make_outer_joins oj.join_root, child
+              if join_type == Arel::Nodes::OuterJoin
+                make_left_outer_joins oj.join_root, child
```

Comment:
This is a question i asked in the initial message of the pull request.:

> The only thing I'm not sure about is if my fix should use make_outer_joins as it was before and as is used in walk or if it should use make_left_outer_joins which is used above in the same method.

Since I'm doing basically the same code as what is done at line 110, I used the same one as is used there. The code of those 2 methods is similar, except for one using `table_aliases_for`, which is very obcsure to me.

So I leave that one last choice up to someone more experienced for those internals.

Note, this has been changed on HEAD, the 2 methods were replaced by a single one, but it has a parameter to specify if aliasing is to be done or not. Once the choice is made, i'll rebase with the chosen way.